### ISSUE TITLE
Convention: private or not private

### ISSUE THREAD

**Issue by Foxandxss** [ng-bootstrap/core#72]
Should we define our component variable as private?

```
class Foo {
  foo;
  bar;
}
```

Or

```
class Foo {
  private foo;
  private bar;
}
```

This doesn't include private functions when we know that they are not mean to be used in public like helper functions.

**Comment by Foxandxss** [comment]
I personally vote for we should not mark them as private. No one seems to be doing that. Also check #73

**Comment by wesleycho** [comment]
I'd say we should only mark stuff as private that we explicitly intend on being private.

**Comment by pkozlowski-opensource** [comment]
> I'd say we should only mark stuff as private that we explicitly intend on being private.

Ha! But the whole notation of being "private" is not that crystal-clear with ng2! What should be considered private?
* things that no-one can access (template, #mydir, sub-class)
* things that a template can access but not #mydir, sub-class?
* etc.

What I'm trying to say is that there are many ways of interacting with components so we need to be clear about what we want to communicate with the `private` keyword.

**Comment by Foxandxss** [comment]
That is correct my friend.

Let's start raising something important in here.

`private` won't be used for end users at all. Imagine this code:

```
@Component({
  selector: 'foo',
  template: 'Foo'
})
class Foo {
  variable = 10;
}
```

A user can do:

```
<foo #foo></foo>
{{foo.variable}}
```

and that will work perfectly. Now mark that `variable` as private in the component and try again. Does it work? Definitely. Marking it as private is not useful in there. Check this [plunker](http://plnkr.co/edit/a1D521y5lQlTUD3vplFK?p=preview).

So my vote is, nothing is private, with one exception. Imagine a progressbar:

```
@Input()
set value(newValue) {
  this._value = newValue;
  this.calculatePercentage();
}

private calculatePercentage() {
  ...
}
```

The private in here is for "Hey, this is simply internal logic that you shouldn't be worried about. Notice I didn't use an underscore for `calculatePercentage`. That is for my an old convention for when we didn't have that private keyword.

So TL;DR, use private for internal logic and leave the rest public.

Thoughts?

**Comment by wesleycho** [comment]
I'm fine with this - there's not much value in using private other than semantics, so this works for me.

**Comment by Foxandxss** [comment]
Let me reopen this, I am seeing that angular for private methods uses both `private` and an underscore:

```
private _foo(...) { }
```

Not sure if we should follow angular conventions in here. Thoughts?

**Comment by pkozlowski-opensource** [comment]
From what I understand from our last discussion we are going to `_`-prefix that is marked as `private`. We would be marking things as `private` only if not used in a teamplate nor meant as extension points through class inheritance.

**Comment by Foxandxss** [comment]
Yes, forgot to close this before. All is on the wiki.
